index.js

```diff
@@ -1,93 +1,134 @@
+   * @param {Object} [info] extra data with wich to tag `error`
+   * @return {Error} the modified `error` argument
+   */
+  static tag(error, message, info) {
+    const oError = /** @type{OError} */ (error)
 
-/**
- * Return the `stack` property from `error` and recursively append the `stack`
- * properties from the error's causes, if any.
- *
- * @param {?Error} error assumed not to have circular causes
- * @return {string}
- */
-function getFullStack(error) {
-  if (!error) return ''
-  const causeStack = getFullStack(error.cause)
-  if (causeStack) return error.stack + '\ncaused by: ' + causeStack
-  return error.stack
+    if (!oError._oErrorTags) oError._oErrorTags = []
+
+    const tag = new TaggedError(message, info)
+
+    // Hide this function in the stack trace.
+    if (Error.captureStackTrace) Error.captureStackTrace(tag, OError.tag)
+
+    oError._oErrorTags.push(tag)
+
+    return error
+  }
+
+  /**
+   * The merged info from any `tag`s on the given error.
+   *
+   * If an info property is repeated, the last one wins.
+   *
+   * @param {Error | null | undefined} error any errror (may or may not be an `OError`)
+   * @return {Object}
+   */
+  static getFullInfo(error) {
+    const info = {}
+
+    if (!error) return info
+
+    const oError = /** @type{OError} */ (error)
+
+    if (typeof oError.info === 'object') Object.assign(info, oError.info)
+
+    if (oError._oErrorTags) {
+      for (const tag of oError._oErrorTags) {
+        Object.assign(info, tag.info)
+      }
+    }
+
+    return info
+  }
+
+  /**
+   * Return the `stack` property from `error`, including the `stack`s for any
+   * tagged errors added with `OError.tag` and for any `cause`s.
+   *
+   * @param {Error | null | undefined} error any error (may or may not be an `OError`)
+   * @return {string}
+   */
+  static getFullStack(error) {
+    if (!error) return ''
+
+    const oError = /** @type{OError} */ (error)
+
+    let stack = oError.stack
+
+    if (oError._oErrorTags) {
+      for (const tag of oError._oErrorTags) {
+        stack += `\n${tag.stack}`
+      }
+    }
+
+    const causeStack = oError.cause && OError.getFullStack(oError.cause)
+    if (causeStack) {
+      stack += '\ncaused by:\n' + indent(causeStack)
+    }
+
+    return stack
+  }
 }
 
 /**
- * Is `error` or one of its causes an instance of `klass`?
+ * Used to record a stack trace every time we tag info onto an Error.
  *
- * @param  {?Error} error assumed not to have circular causes
- * @param  {function} klass
- * @return {Boolean}
+ * @private
+ * @extends OError
  */
-function hasCauseInstanceOf(error, klass) {
-  if (!error) return false
-  return error instanceof klass || hasCauseInstanceOf(error.cause, klass)
-}
+class TaggedError extends OError {}
 
-OError.getFullInfo = getFullInfo
-OError.getFullStack = getFullStack
-OError.hasCauseInstanceOf = hasCauseInstanceOf
+function indent(string) {
+  return string.replace(/^/gm, '    ')
```

Comment:
Why not use `left-pad` 😆?

But seriously, there is [`.padLeft()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart) now.

Reply:
I appreciate the sentiment... but it looks like `padStart` takes an expected length, which is not quite what we want here --- we just want to indent every line with some spaces. So, I'm not seeing an easy way of doing this with `padStart`.